src/compose/application-manager.ts

```diff
@@ -0,0 +1,845 @@
+			(svc) =>
+				_.find(availableImages, {
+					dockerImageId: svc.config.image,
+					imageId: svc.imageId,
+				}) ?? _.find(availableImages, { dockerImageId: svc.config.image }),
+		),
+	) as imageManager.Image[];
+	const targetImages = _.flatMap(target, (app) =>
+		_.map(app.services, imageForService),
+	);
+
+	const availableAndUnused = _.filter(
+		availableImages,
+		(image) =>
+			!_.some(
+				currentImages.concat(targetImages),
+				(imageInUse) =>
+					imageManager.isSameImage(image, imageInUse) ||
+					image.id === imageInUse.id ||
+					image.dockerImageId === imageInUse.dockerImageId,
+			),
+	);
+
+	const imagesToDownload = _.filter(
+		targetImages,
+		(targetImage) =>
+			!_.some(availableImages, (available) =>
+				imageManager.isSameImage(available, targetImage),
+			),
+	);
+
+	const targetImageDockerIds = _.fromPairs(
+		_.flatMap(target, allImageDockerIdsForTargetApp),
+	);
+
+	// Images that are available but we don't have them in the DB with the exact metadata:
+	let imagesToSave: imageManager.Image[] = [];
+	if (!localMode) {
+		imagesToSave = _.filter(targetImages, (targetImage) => {
+			const isActuallyAvailable = _.some(availableImages, (availableImage) => {
+				if (imageManager.isSameImage(availableImage, targetImage)) {
+					return true;
+				}
+				if (
+					availableImage.dockerImageId ===
+					targetImageDockerIds[targetImage.name]
+				) {
+					return true;
+				}
+				return false;
+			});
+			const isNotSaved = !_.some(availableWithoutIds, (img) =>
+				_.isEqual(img, targetImage),
+			);
+			return isActuallyAvailable && isNotSaved;
+		});
+	}
+
+	const deltaSources = _.map(imagesToDownload, (image) => {
+		return bestDeltaSource(image, availableImages);
+	});
+	const proxyvisorImages = proxyvisor.imagesInUse(current, target);
+
+	const potentialDeleteThenDownload = _(current)
+		.flatMap((app) => _.values(app.services))
+		.filter(
+			(svc) =>
+				svc.config.labels['io.balena.update.strategy'] ===
+					'delete-then-download' && svc.status === 'Stopped',
+		)
+		.value();
+
+	const imagesToRemove = _.filter(
+		availableAndUnused.concat(potentialDeleteThenDownload.map(imageForService)),
+		(image) => {
+			const notUsedForDelta = !_.includes(deltaSources, image.name);
+			const notUsedByProxyvisor = !_.some(proxyvisorImages, (proxyvisorImage) =>
+				imageManager.isSameImage(image, { name: proxyvisorImage }),
+			);
+			return notUsedForDelta && notUsedByProxyvisor;
+		},
+	);
+
+	return imagesToSave
+		.map((image) => ({ action: 'saveImage', image } as CompositionStep))
+		.concat(imagesToRemove.map((image) => ({ action: 'removeImage', image })));
+}
+
+async function getAppContainerIds(currentApps: InstancedAppState) {
+	const containerIds: { [appId: number]: Dictionary<string> } = {};
+	await Promise.all([
+		_.map(currentApps, async (_app, appId) => {
+			const intAppId = parseInt(appId, 10);
+			containerIds[intAppId] = await serviceManager.getContainerIdMap(intAppId);
+		}),
+	]);
```

Comment:
This looks broken to me, is it meant to be
```suggestion
	await Promise.all(
		_.map(currentApps, async (_app, appId) => {
			const intAppId = parseInt(appId, 10);
			containerIds[intAppId] = await serviceManager.getContainerIdMap(intAppId);
		}),
	);
```
?